ansible/roles/etcd/tasks/main.yml

```diff
@@ -1,4 +1,106 @@
 ---
+### Create TLS assets using kubeadm
+- name: download standalone kubeadm for cert generation
+  get_url:
+    url: "{{ kubeadm_release_url }}"
+    dest: /usr/local/bin/kubeadm
+    mode: 0744
+    force: no
+
+- name: determine whether kubeadm needs to create etcd ca
+  stat:
+    path: "/etc/kubernetes/pki/etcd/{{ item }}"
+  with_items:
+    - ca.crt
+    - ca.key
+  register: etcd_ca
+
+- name: determine whether kubeadm needs to create etcd peer and server certs
+  stat:
+    path: "/etc/kubernetes/pki/etcd/{{ item }}"
+  with_items:
+    - peer.crt
+    - peer.key
+    - server.crt
+    - server.key
+  register: etcd_local_tls_assets
+
+- name: determine whether kubeadm needs to create the apiserver etcd client certs
+  stat:
+    path: "/etc/kubernetes/pki/{{ item }}"
+  with_items:
+    - apiserver-etcd-client.crt
+    - apiserver-etcd-client.key
+  register: apiserver_etcd_certs
+  delegate_to: "{{ groups['etcd'][0] }}"
+
+- name: create /etc/kubernetes directory
+  file:
+    dest: /etc/kubernetes
+    state: directory
+    owner: root
+    group: root
+
+- name: laydown etcd.conf to be used for cert generation
+  template:
+    src: etc/kubernetes/etcd.conf
+    dest: /etc/kubernetes/etcd.conf
+  with_items: "{{ etcd_ca.results }}"
+  when: item.stat.exists == False
+
+- name: generate etcd ca certs with kubeadm
+  command: "/usr/local/bin/kubeadm alpha phase certs etcd-ca"
+  delegate_to: "{{ groups['etcd'][0] }}"
+  run_once: True
+  with_items: "{{ etcd_ca.results }}"
+  when: item.stat.exists == False
+
+- name: generate apiserver client certs with kubeadm
+  command: "/usr/local/bin/kubeadm alpha phase certs apiserver-etcd-client"
+  delegate_to: "{{ groups['etcd'][0] }}"
+  run_once: True
+  with_items: "{{ apiserver_etcd_certs.results }}"
+  when: item.stat.exists == False
+
+- name: slurp the etcd ca certs
+  slurp:
+    src: /etc/kubernetes/pki/etcd/{{ item }}
+  with_items:
+    - ca.crt
+    - ca.key
+  register: etcd_ca_pki
+  delegate_to: "{{ groups['etcd'][0] }}"
```

Comment:
I see the `[0]` in a few places here and `|first` later on, should these be consistent?

Reply:
Good call. I've updated to `|first`